### PR TITLE
Collapse IdleDeadline's deadline and isExceeded into IdleDeadline.timeRemaining()

### PR DESCRIPTION
**Pull request by rmcilroy**
Replace IdleDeadline.deadline and IdleDeadline.isExceeded() with IdleDeadline.timeRemaining()

Simplifies the API as discussed in #11

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -227,10 +227,8 @@
 };
 
 interface IdleDeadline {
-  readonly attribute DOMHighResTimeStamp deadline;
+  double timeRemaining();
```
Comment (igrigorik) [inline index.html]: Should this still be DOMHighResTimeStamp?

Comment (rmcilroy) [inline index.html]: I thought of that, but according to the DOMHighResTimeStamp spec it is "a time value measured relative to the navigationStart", which is not the case for this return value (this return value is a time delta, not an absolute time).

Comment (igrigorik) [inline index.html]: I think you may be looking at an outdated definition.

"The DOMHighResTimeStamp type is used to store a time value measured relative from the time origin _or a time value that represents a duration between two DOMHighResTimeStamps_." - [source](http://w3c.github.io/hr-time/#dom-domhighrestimestamp)

**Comment by rmcilroy** [comment]
Ahh right, in that case sounds good. Updated to a DOMHighResolutionTimestamp.

**Comment by igrigorik** [comment]
LGTM. @toddreifsteck @eliperelman any objections?

**Comment by rmcilroy** [comment]
I made a couple of additional changes in https://github.com/rmcilroy/requestidlecallback/commit/ca61eebda7183b6e3014d0b473a63f4c1327e1d6 based on feedback from @esprehn:
- Made timeRemaining an attribute rather than a method
- Made timeRemaining not return negative values (clamped at 0 when the deadline passes).

PTAL, thanks.

**Comment by igrigorik** [comment]
Looks good!

**Comment by toddreifsteck** [comment]
LGTM. @eliperelman, you ok with this?

**Comment by eliperelman** [comment]
+1 from me!

**Comment by igrigorik** [comment]
Then let it be so! Thanks guys :)

**Comment by igrigorik** [comment]
> Why did you replace the isExceeded() method with the timeRemaining attribute? That makes the implementation impossible to polyfill.﻿[source](https://plus.google.com/u/0/+IlyaGrigorik/posts/bPNjgMwcMKs)

That's a gotcha I didn't consider. Should we go back to a method?

**Comment by rmcilroy** [comment]
You should be able to use a getter to implement the timeRemaining attribute in a pollyfill. For example, something like:

function IdleCallbackDeadline(deadline) {
  var idleDeadline = {
    get timeRemaining() { return deadline - performance.now(); }
  };
  return idleDeadline;
};
var deadline = IdleCallbackDeadline(performance.now() + 10000);
console.log(deadline.timeRemaining);

So I don't think it is necessary to go back to a method to make it possible to polyfill.

**Comment by joshduck** [comment]
Using getters won't work for IE8. Globally that browser doesn't have a huge market share, but in some niches they are still important; which is a consideration when using requestIdleCallback in a framework or library.

Using a method would mirror the Date.now() or performance.now() functions. And I'd argue that a method is more appropriate for a value that is expected to change. Otherwise you end up with code that looks like it can safely be refactored like this:

  if (deadline.timeRemaining > 0) doFoo();
  if (deadline.timeRemaining > 0) doBar();

**Comment by rmcilroy** [comment]
I didn't realize IE8 doesn't support getters. However, it doesn't have requestAnimationFrame or performance.now() either, so I'm not sure how you would polyfill it on IE8 in such a way as to get any reasonable value for the timeRemaining attribute either way. What did you have in mind for an IE8 polyfill? 

Maybe it would be better just to use setTimeout to polyfill and have the callbacks be able to deal with the deadline argumnent possibly being undefined, rather than trying to create an IdleCallbackDeadline object which might not accurately reflect deadlines?

**Comment by joshduck** [comment]
Having to handle `undefined` in all client code would be confusing. The callback would have to have it's own logic for continuing or terminating execution, and this would only be executed for some clients, leading to buggy code.

```
function doWork(deadline) {
  var iterations = 10;
  while (deadline.timeRemaining === undefined ? iterations-- : deadline.timeRemaining > 0) 
    doSomethingExpensive();
}
```

If you had a `timeRemaining()` or `isExceeded()` method you could easily build a mock to pass into the doWork method. This would be useful for frameworks or libraries that do their own scheduling.

```
 class FakeIdleDeadline {
    constructor(iterations) { this.iterations = iterations }
    timeRemaining() { return this.iterations--; }
 }   
 doWork(new FakeIdleDeadline(10));
```

Or, if you only cared about millisecond granularity you could use `Date.now()`.

**Comment by rmcilroy** [comment]
Ok sounds reasonable. I've uploaded a change to the spec in pull request #14. PTAL, thanks.

**Comment by igrigorik** [comment]
@joshduck resolved in https://github.com/w3c/requestidlecallback/pull/14 - thanks for your feedback.

**Pull request merged**